agent/consul/stream/event.go

```diff
@@ -64,6 +72,20 @@ func (e Event) Filter(f func(Event) bool) (Event, bool) {
 	return e, true
 }
 
+// PayloadEvents is an Payload which contains multiple Events.
+type PayloadEvents []Event
+
+// TODO: this method is not called, but needs to exist so that we can store
+// a slice of events as a payload. In the future we should be able to refactor
+// Event.Filter so that this FilterByKey includes the re-slicing.
```

Comment:
Is this an argument for changing the `FilterByKey` method now so that it isn't just a predecate but actually returns a `Payload`. That way even composite types like this one could do something meaningful here as they could apply the filtering to their list and return a new list.

Regular event could just do the logic they do now and either return themselves or nil of they should be filtered?

Reply:
I hadn't considered returning a `Payload` from `FilterByKey`. I think that's an excellent idea and solves part of the problem.

This TODO is missing a bit of context. One of the challenges here is that `Event.Filter` is used in two places, once for filtering by key+namespace, and once for filtering ACLs.  In order to use it for both we'll need another method on the `Payload` interface to handle ACL filtering, and I decided to hold off on adding that as part of this change.

I think with your suggestion this is now easier to resolve. There are a number of test changes in #9073 which is a follow up to this PR and which will conflict pretty heavily if I resolve this TODO here.

I'll merge these 2 PRs first and resolve this TODO and the `FilterByKey` naming in a follow up.